File: src/services/isbn.py
import re
import unicodedata

from src.core.errors import ValidationError


def _slugify_identifier(text: str) -> str:
    normalized = unicodedata.normalize("NFD", text)
    without_accents = "".join(ch for ch in normalized if unicodedata.category(ch) != "Mn")
    slug = re.sub(r"[^a-zA-Z0-9]+", "-", without_accents).strip("-").lower()
    return slug or "livro"


def resolve_book_key(raw: str | None, filename: str | None = None) -> str:
    """ISBN opcional: usa campo, nome do arquivo (se for ISBN) ou slug do arquivo."""
    if raw and raw.strip():
        cleaned = re.sub(r"[^0-9Xx]", "", raw.strip())
        if cleaned:
            return normalize_isbn(raw)

    if filename:
        stem = (filename or "").rsplit(".", 1)[0].strip()
        if stem:
            try:
                return normalize_isbn(stem)
            except ValidationError:
                return _slugify_identifier(stem)

    raise ValidationError("Informe um ISBN ou use um arquivo PDF com nome identificável.")


_SLUG_PATTERN = re.compile(r"^[a-z0-9](?:[a-z0-9-]*[a-z0-9])?$")


def validate_book_key(raw: str) -> str:
    """Aceita ISBN (10/13) ou slug gerado a partir do nome do arquivo."""
    cleaned = (raw or "").strip()
    if not cleaned:
        raise ValidationError("Identificador do livro obrigatorio.")
    if len(cleaned) > 128:
        raise ValidationError("Identificador do livro muito longo (max. 128 caracteres).")

    digits = re.sub(r"[^0-9Xx]", "", cleaned)
    if len(digits) in (10, 13):
        try:
            return normalize_isbn(cleaned)
        except ValidationError:
            pass

    if _SLUG_PATTERN.fullmatch(cleaned):
        return cleaned

    raise ValidationError("Identificador do livro invalido.")


def normalize_isbn(raw: str) -> str:
    cleaned = re.sub(r"[^0-9Xx]", "", (raw or "").strip())
    if len(cleaned) == 10:
        if not _is_valid_isbn10(cleaned):
            raise ValidationError("ISBN-10 invalido.")
        return cleaned.upper()
    if len(cleaned) == 13:
        if not _is_valid_isbn13(cleaned):
            raise ValidationError("ISBN-13 invalido.")
        return cleaned
    raise ValidationError("ISBN deve ter 10 ou 13 caracteres validos.")


def _is_valid_isbn10(isbn: str) -> bool:
    if not re.match(r"^\d{9}[\dXx]$", isbn):
        return False
    total = 0
    for idx, ch in enumerate(isbn[:9], start=1):
        total += idx * int(ch)
    check = isbn[9].upper()
    total += 10 * (10 if check == "X" else int(check))
    return total % 11 == 0


def _is_valid_isbn13(isbn: str) -> bool:
    if not re.match(r"^\d{13}$", isbn):
        return False
    total = 0
    for idx, ch in enumerate(isbn[:12]):
        weight = 1 if idx % 2 == 0 else 3
        total += int(ch) * weight
    check_digit = (10 - (total % 10)) % 10
    return check_digit == int(isbn[-1])

File: lib/isbn.ts
export function normalizeIsbn(value: string): string {
  return value.replace(/[^0-9Xx]/g, "").toUpperCase();
}

export function isValidIsbn10(raw: string): boolean {
  const isbn = normalizeIsbn(raw);
  if (!/^\d{9}[\dX]$/.test(isbn)) return false;
  const sum = isbn.split("").reduce((acc, char, index) => {
    const value = char === "X" ? 10 : Number(char);
    return acc + value * (10 - index);
  }, 0);
  return sum % 11 === 0;
}

export function isValidIsbn13(raw: string): boolean {
  const isbn = normalizeIsbn(raw);
  if (!/^\d{13}$/.test(isbn)) return false;
  const sum = isbn
    .slice(0, 12)
    .split("")
    .reduce((acc, char, index) => acc + Number(char) * (index % 2 === 0 ? 1 : 3), 0);
  const checkDigit = (10 - (sum % 10)) % 10;
  return checkDigit === Number(isbn[12]);
}

export function isValidIsbn(raw: string): boolean {
  return isValidIsbn10(raw) || isValidIsbn13(raw);
}

export function extractIsbnFromFilename(filename: string): string | null {
  const base = filename.replace(/\.pdf$/i, "").trim();
  if (!/^\d{10}$|^\d{13}$/.test(base)) return null;
  return base;
}

/** Identificador do livro: ISBN informado, ISBN no nome do arquivo ou slug do arquivo. */
export function resolveBookKey(isbnInput: string, filename: string): string {
  const trimmed = isbnInput.trim();
  if (trimmed) {
    const normalized = normalizeIsbn(trimmed);
    if (isValidIsbn(normalized)) return normalized;
  }

  const fromFile = extractIsbnFromFilename(filename);
  if (fromFile && isValidIsbn(fromFile)) return fromFile;

  const stem = filename.replace(/\.pdf$/i, "").trim();
  const slug = stem
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "");

  return slug || "livro";
}
